src/Node/Node.php

```diff
@@ -0,0 +1,263 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file was originally part of the league/commonmark package.
+ *
+ * (c) Colin O'Dell <colinodell@gmail.com>
+ *
+ * Original code based on the CommonMark JS reference parser (https://bitly.com/commonmark-js)
+ *  - (c) John MacFarlane
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace UnicornFail\Emoji\Node;
+
+abstract class Node
+{
+    /**
+     * @var int
+     *
+     * @psalm-readonly-allow-private-mutation
+     */
+    protected $depth = 0;
+
+    /**
+     * @var Node|null
+     *
+     * @psalm-readonly-allow-private-mutation
+     */
+    protected $parent;
```

Comment:
I think the reuse of commonmark's full AST implementation might be overkill here.  In commonmark, we need a way to represent a **nested tree** of parsed objects, and that representation needed to be **implemented as a multi-level, doubly-linked list** so we can iterate it without using recursion.  But in this case, I think we could get by with a **flat, non-tree** representation since we're dealing with a stream of characters.  This could be a linked list, doubly-linked list, or even just a simple `iterable` of nodes/tokens. I just don't think we really need the concept of parents and children.

Reply:
And if we're making it flat I also think we can remove the concept of block vs inline nodes.  The object that contains the nodes doesn't necessarily have to be a node itself in this case.

Reply:
I agree that the "nested/depth" aspect of this isn't necessary here; this was just a literal c&p for times sake.

I mostly extracted it because I would like to keep the node traversal and injection methods (i.e. append, detach, insertBefore/After, previous, next, etc.). These kinds of methods make it easier to manipulate the AST, regardless of its depth.

I do think there should be a simple (flat) `NodeInterface` and then a `HierarchicalNodeInterface` perhaps? Again, the main goal here is to see what aspects could be extracted out that aren't really "commonmark" specific and could be reused between the two projects and provide a common interface for both to use so what is parsed here could be used "as is" in commonmark without having to do unnecessary and (ultimately) costly object transforms.

Reply:
This has been simplified/removed.